iOS/UI/Swipeable Input/SwipeInputPanGestureHandler.swift

```diff
@@ -0,0 +1,203 @@
+//
+//  SwipeInputPanGestureRecognizer.swift
+//  Jibber
+//
+//  Created by Martin Young on 1/26/22.
+//  Copyright © 2022 Benjamin Dodgson. All rights reserved.
+//
+
+import Foundation
+import Lottie
+
+class SwipeInputPanGestureHandler {
+
+    let inputView: SwipeableInputAccessoryView
+
+    init(inputView: SwipeableInputAccessoryView) {
+        self.inputView = inputView
+    }
+
+    /// An object to give the user touch feedback when performing certain actions.
+    private var impactFeedback = UIImpactFeedbackGenerator(style: .rigid)
+    private var previewView: PreviewMessageView?
+    /// The center point of the preview view when the pan started.
+    private var initialPreviewCenter: CGPoint?
+    /// How far the preview view can be dragged left or right.
+    private let maxXOffset: CGFloat = 40
+    /// If true, the preview view is currently in the drop zone.
+    private var isPreviewInDropZone = false
+
+    func handle(pan: UIPanGestureRecognizer) {
+        guard self.shouldHandlePan() else { return }
+
+        let panOffset = pan.translation(in: nil)
+
+        switch pan.state {
+        case .possible:
+            break
+        case .began:
+            self.handlePanBegan()
+        case .changed:
+            self.handlePanChanged(withOffset: panOffset)
+        case .ended:
+            self.handlePanEnded(withOffset: panOffset)
+        case .cancelled, .failed:
+            self.handlePanFailed()
+        @unknown default:
+            break
+        }
+    }
+
+    private func shouldHandlePan() -> Bool {
+        // Only handle pans if the user has input a sendable message.
+        let object = SendableObject(kind: self.inputView.currentMessageKind,
+                                    context: self.inputView.currentContext,
+                                    emotion: self.inputView.currentEmotion,
+                                    previousMessage: self.inputView.editableMessage)
+        return object.isSendable
+    }
+
+    private func handlePanBegan() {
+        let object = SendableObject(kind: self.inputView.currentMessageKind,
+                                    context: self.inputView.currentContext,
+                                    emotion: self.inputView.currentEmotion,
+                                    previousMessage: self.inputView.editableMessage)
+        self.inputView.sendable = object
+
+        // Hide the input area. The preview view will take its place during the pan.
+        self.inputView.inputContainerView.alpha = 0
+
+        // Stop any swipe hint animations if they're playing.
+        self.inputView.updateSwipeHint(shouldPlay: false)
```

Comment:
The one functional addition is that we cancel the hint animation if the user starts to swipe.